File: DropDownTableViewObjC/DropDownTableViewObjC/AnotherTestController.m
//
//  AnotherTestController.m
//  DropDownTableViewObjC
//
//  Created by NSSimpleApps on 15.01.17.
//  Copyright © 2017 NSSimpleApps. All rights reserved.
//

#import "AnotherTestController.h"

@interface AnotherTestController ()

@property (assign) NSInteger numberOfSubrows;
@property (assign) NSInteger numberOfRows;

@end

@implementation AnotherTestController

- (void)viewDidLoad {
    
    [super viewDidLoad];
    
    self.numberOfSubrows = 3;
    self.numberOfRows = 10;
    
    self.title = @"Another test";
    self.navigationItem.rightBarButtonItems =
    @[
    [[UIBarButtonItem alloc] initWithTitle:@"Insert"
                                     style:UIBarButtonItemStylePlain
                                    target:self
                                    action:@selector(insertRows:)],
    [[UIBarButtonItem alloc] initWithTitle:@"Delete"
                                     style:UIBarButtonItemStylePlain
                                    target:self
                                    action:@selector(deleteRows:)]
    ];
}

- (void)insertRows:(UIBarButtonItem *)sender {
    
    self.numberOfRows += 2;
    [self.tableView insertRowsAtRows:@[@0, @1] withRowAnimation:UITableViewRowAnimationAutomatic];
}

- (void)deleteRows:(UIBarButtonItem *)sender {
    
    self.numberOfRows -= 2;
    [self.tableView deleteRowsAtRows:@[@0, @1] withRowAnimation:UITableViewRowAnimationAutomatic];
}

- (NSInteger)numberOfRowsInTableView:(UITableView *)tableView {
    
    return self.numberOfRows;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfSubrowsInRow:(NSInteger)row {
    
    NSLog(@"%ld", (long)self.numberOfSubrows);
    
    return self.numberOfSubrows;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRow:(NSInteger)row indexPath:(NSIndexPath *)indexPath {
    
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"Cell" forIndexPath:indexPath];
    cell.textLabel.text = [NSString stringWithFormat:@"Row %ld", (long)row];
    cell.detailTextLabel.text = nil;
    
    if (self.nsk_selectedRow) {
        
        NSInteger selectedRow = self.nsk_selectedRow.integerValue;
        
        if (row == selectedRow) {
            
            cell.accessoryView = [[UIImageView alloc] initWithImage:[UIImage imageNamed:@"up_arrow"]];
            
        } else {
            
            cell.accessoryView = [[UIImageView alloc] initWithImage:[UIImage imageNamed:@"down_arrow"]];
        }
    }
    
    return cell;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForSubrow:(NSInteger)subrow inRow:(NSInteger)row indexPath:(NSIndexPath *)indexPath {
    
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"Cell" forIndexPath:indexPath];
    cell.textLabel.text = [NSString stringWithFormat:@"Subrow %ld", (long)row];
    cell.detailTextLabel.text = @(subrow).stringValue;
    cell.accessoryView = nil;
    
    
    
    return cell;
}

- (void)tableView:(UITableView *)tableView didSelectRow:(NSInteger)row {
    
    if (self.nsk_selectedRow) {
        
        NSInteger selectedRow = self.nsk_selectedRow.integerValue;
        
        if (row == selectedRow) {
            
            [tableView cellForRowAtRow:row].accessoryView = [[UIImageView alloc] initWithImage:[UIImage imageNamed:@"down_arrow"]];
            [tableView deselectRow:row animated:YES];
            
        } else {
            
            [tableView cellForRowAtRow:row].accessoryView = [[UIImageView alloc] initWithImage:[UIImage imageNamed:@"up_arrow"]];
            [tableView cellForRowAtRow:selectedRow].accessoryView = [[UIImageView alloc] initWithImage:[UIImage imageNamed:@"down_arrow"]];
        }
        
    } else {
        
        [tableView cellForRowAtRow:row].accessoryView = [[UIImageView alloc] initWithImage:[UIImage imageNamed:@"up_arrow"]];
    }
    
    [super tableView:tableView didSelectRow:row];
}

- (void)tableView:(UITableView *)tableView didSelectSubrow:(NSInteger)subrow inRow:(NSInteger)row {
    
    NSLog(@"%s, %ld, %ld", __func__, (long)subrow, (long)row);
}

- (void)tableView:(UITableView *)tableView didDeselectRow:(NSInteger)row {
    
    NSLog(@"%s, %ld", __func__, (long)row);
}

- (void)tableView:(UITableView *)tableView didDeselectSubrow:(NSInteger)subrow inRow:(NSInteger)row {
    
    NSLog(@"%s, %ld, %ld", __func__, (long)subrow, (long)row);
}

@end

File: DropDownTableViewSwift/DropDownTableViewSwift/AnotherTestController.swift
//
//  TestViewController.swift
//  DropDownTableView
//
//  Created by NSSimpleApps on 26.12.15.
//  Copyright © 2015 NSSimpleApps. All rights reserved.
//

import UIKit

class AnotherTestController: DropDownTableViewController {
    
    fileprivate var numberOfSubrows = 3
    fileprivate var numberOfRows = 10
    
    override func viewDidLoad() {
        
        super.viewDidLoad()
        
        self.title = "Another test"
        
        self.navigationItem.rightBarButtonItems =
        [UIBarButtonItem(title: "Insert", style: .plain, target: self, action: #selector(self.insertRows(_:))),
        UIBarButtonItem(title: "Delete", style: .plain, target: self, action: #selector(self.deleteRows(_:)))]
    }
    
    func insertRows(_ sender: UIBarButtonItem) {
        
        self.numberOfRows += 2
        
        self.tableView.insertRows(at: [0, 1], with: .automatic)
    }
    
    func deleteRows(_ sender: UIBarButtonItem) {
        
        self.numberOfRows -= 2
        
        self.tableView.deleteRows(at: [0, 1], with: .automatic)
    }
    
    override func numberOfRows(in tableView: UITableView) -> Int {
        
        return self.numberOfRows
    }
    
    override func tableView(_ tableView: UITableView, numberOfSubrowsInRow row: Int) -> Int {
        
        return self.numberOfSubrows
    }
    
    override func tableView(_ tableView: UITableView, cellForRow row: Int, indexPath: IndexPath) -> UITableViewCell {
        
        let cell = tableView.dequeueReusableCell(withIdentifier: "Cell", for: indexPath)
        cell.textLabel?.text = "Row " + String(row)
        cell.detailTextLabel?.text = nil
        
        if row == self.nsk_selectedRow {
            
            cell.accessoryView = UIImageView(image: UIImage(named: "up_arrow"))
            
        } else {
            
            cell.accessoryView = UIImageView(image: UIImage(named: "down_arrow"))
        }
        
        return cell
    }
    
    override func tableView(_ tableView: UITableView, cellForSubrow subrow: Int, inRow row: Int, indexPath: IndexPath) -> UITableViewCell {
        
        let cell = tableView.dequeueReusableCell(withIdentifier: "Cell", for: indexPath)
        cell.textLabel?.text = "Subrow " + String(row)
        cell.detailTextLabel?.text = String(subrow)
        cell.accessoryView = nil
        
        return cell
    }
    
    override func tableView(_ tableView: UITableView, didSelectRow row: Int) {
        
        switch (self.nsk_selectedRow, row) {
            
        case (let sr?, _) where row == sr:
            tableView.cellForRow(at: row)?.accessoryView =  UIImageView(image: UIImage(named: "down_arrow"))
            tableView.deselect(row: row, animated: true)
            break
            
        case (let sr?, _) where row != sr:
            tableView.cellForRow(at: row)?.accessoryView = UIImageView(image: UIImage(named: "up_arrow"))
            tableView.cellForRow(at: sr)?.accessoryView = UIImageView(image: UIImage(named: "down_arrow"))
            break
            
        case (nil, _):
            tableView.cellForRow(at: row)?.accessoryView = UIImageView(image: UIImage(named: "up_arrow"))
            break
            
        default:
            break
        }
        
        super.tableView(tableView, didSelectRow: row)
    }
    
    override func tableView(_ tableView: UITableView, didSelectSubrow subrow: Int, inRow row: Int) {
        
        print(#function, subrow, row)
    }
    
    override func tableView(_ tableView: UITableView, didDeselectRow row: Int) {
        
        print(#function, row)
    }
    
    override func tableView(_ tableView: UITableView, didDeselectSubrow subrow: Int, inRow row: Int) {
        
        print(#function, subrow, row)
    }
}

